## Mutual Information, Self-information and Split Learning

In this exercise we want to see how split-learning is helping privacy by decreasing the information content in a raw input. This is a simplified example, just to help the understanding of the concepts in this lesson. 

We will be using MNIST data for this exercise. We will first download and load the data. Then, we will load a pretrained small DNN. Our aim is to compare the information in the raw inputs of the MNIST test set, with the information in the output of the final convolution layer, to see if there is information degredation.

In [1]:
import torch
from torch import nn

from lenet_5 import LeNet5_5
from torchvision.datasets.mnist import MNIST
import torchvision.transforms as transforms
from torch.utils.data import DataLoader
import numpy as np

#### Load data

In [2]:
BATCH_SIZE = 256
BATCH_TEST_SIZE = 1024
data_train = MNIST('/Users/turgutlu/.syft/data/',
                   download=False,
                   transform=transforms.Compose([
                       transforms.Resize((32, 32)),
                       transforms.ToTensor()]))
data_test = MNIST('/Users/turgutlu/.syft/data/',
                  train=False,
                  download=False,
                  transform=transforms.Compose([
                      transforms.Resize((32, 32)),
                      transforms.ToTensor()]))
data_train_loader = DataLoader(data_train, batch_size = BATCH_SIZE , shuffle=True, num_workers=8)
data_test_loader = DataLoader(data_test,  batch_size = BATCH_TEST_SIZE, num_workers=8)
data_test_loader2 = DataLoader(data_test,  batch_size = 1, num_workers=0)

TRAIN_SIZE = len(data_train_loader.dataset)
TEST_SIZE = len(data_test_loader.dataset)
NUM_BATCHES = len(data_train_loader)
NUM_TEST_BATCHES = len(data_test_loader)

#### Load pre-trained model

In [3]:
model_loaded = LeNet5_5()
model_loaded.load_state_dict(torch.load("./LeNet-saved-5"))
criterion = nn.NLLLoss()

#### Validate

In [4]:
def validate (net, criterion):
    net.eval()
    total_correct = 0
    avg_loss = 0.0
    for i, (images, labels) in enumerate(data_test_loader):
        labels = (labels > 5).long()
        output = net(images)
        avg_loss += criterion(output, labels).sum()
        pred = output.detach().max(1)[1]
        total_correct += pred.eq(labels.view_as(pred)).sum()

    avg_loss /= len(data_test)
    print('Test Avg. Loss: %f, Accuracy: %f' % (avg_loss.detach().cpu().item(), float(total_correct) / len(data_test)))
    return 

#### Run validate to check the accuracy of the pretrained model

In [5]:
validate (model_loaded, criterion)

Test Avg. Loss: 0.000023, Accuracy: 0.992900


In [6]:
model_loaded

LeNet5_5(
  (convnet): Sequential(
    (c1): Conv2d(1, 6, kernel_size=(5, 5), stride=(1, 1))
    (relu1): ReLU()
    (s2): MaxPool2d(kernel_size=(2, 2), stride=2, padding=0, dilation=1, ceil_mode=False)
    (c3): Conv2d(6, 16, kernel_size=(5, 5), stride=(1, 1))
    (relu3): ReLU()
    (s4): MaxPool2d(kernel_size=(2, 2), stride=2, padding=0, dilation=1, ceil_mode=False)
    (c5): Conv2d(16, 120, kernel_size=(5, 5), stride=(1, 1))
    (relu5): ReLU()
  )
  (fc): Sequential(
    (f6): Linear(in_features=120, out_features=84, bias=True)
    (relu6): ReLU()
    (f7): Linear(in_features=84, out_features=2, bias=True)
    (sig7): LogSoftmax(dim=-1)
  )
)

### Splitting and measuring information content

At this point, we want to split the network to two parts, and observe how different the information content of the original images and the intermediate activations are. We chose the last convolution layer of the pre-trained model we had as the splitting point. We will feed all the test data to the convolutions, and save their outputs so that we can later use them to quantitatively measure the bits of information. 

#### Save the raw images and the intermediate activations

In [7]:
from tqdm.notebook import tqdm

In [8]:
imgs = []
intermediate_activations = []
total_correct = 0

model_loaded.eval()
with torch.no_grad():
    for i, (images, labels) in tqdm(enumerate(data_test_loader2), total=len(data_test_loader2)):
        imgs.append(images.squeeze(0).view(1,-1))

        x = model_loaded.convnet(images)
        intermediate_activations.append(x.view(1,120))

    np.save("images", torch.cat(imgs).numpy())
    np.save("intermediate_act", torch.cat(intermediate_activations).numpy())

#### Load the Information Toolbox

We'll be using the Information toolbox to calculate mutual information. Let's load it here. If the `ite-repo` folder isn't in the same directory where you're running this notebook, change the file path to the correct location below.

In [10]:
import sys
sys.path.insert(1,'./ite-repo')
import ite

Then we'll load the raw images and intermediate activations as Numpy arrays.

In [11]:
images_raw=np.load("images.npy")
print(images_raw.shape)
intermediate_activation=np.load("intermediate_act.npy")
print(intermediate_activation.shape)

(10000, 1024)
(10000, 120)


#### Mutual Information function

We will calculate self-information and mutual information of the first 1,000 test images. You can remove the slicing to calculate on full 10,000 test images but it will take longer to run.

$$
I(X;Y) = \sum\limits_{(x,y) \in \mathcal{X}\times\mathcal{Y}} P_{XY}(x,y) \log \frac{P_{XY}(x,y)}{P_X(x)P_Y(y)}
$$

In [12]:
co = ite.cost.MIShannon_DKL()

Here we'll calculate the self-information of the raw images.

In [13]:
ds = np.array([1024, 1024])
y = np.concatenate((images_raw[:1000], images_raw[:1000]),axis=1)
print(y.shape)
i = co.estimation(y, ds) 
print(i)

(1000, 2048)
455.97089317698345


Then we'll calculate the mutual information between the raw images and the intermediate activations.

In [14]:
ds = np.array([1024, 120])
y = np.concatenate((images_raw[:1000], intermediate_activation[:1000]),axis=1)
print(y.shape)
i = co.estimation(y, ds) 
print(i)

(1000, 1144)
163.148586532957


We can see that the raw image contained 455 bits of self-information, whereas the intermediate activations only contain 163 bits of information that was originally in the raw image (the 455 bits). This shows that the first layers of the neural network, alone, have degraded more than half of the original information in the raw input. 